design/external-references.md

```diff
@@ -0,0 +1,394 @@
+
+Since AWS CloudFormation has a limit on export name length, and we wouldn't want to restrict the serialization depth, the import/export serializer should trim the name and add a hash of the full name, but only if the total length exceeds the limit.
+
+### Synthesis-time Imports (REQ8)
+
+As mentioned in the previous section, `importObject` will support both synthesis and deploy-time imports by export name.
+
+In order to implement synthesis-time imports, we will add a new environmental context provider to the toolkit which will be able to retrieve a value for a certain CloudFormation named export.
+
+The implementation of this provider will use the CloudFormation ListExports operation to find the exports needed and pass their values in through the CDK context mechanism.
+
+Bear in mind that once a value has been retrieved, it will automatically be saved in the local `cdk.json` and won't be retrieved again until `cdk context --reset` is called.
+
+Since synthesis-time resolution doesn't create strong coupling between the stacks at the CloudFormation
+level, production and operational issues can arise if the producing stack deletes an exported resource. On the other hand, we hear from customers that strong-referencing behavior of `Fn::ImportValue` is sometimes a curse. Customers tell us that they found themselves stuck with unremovable or stacks that cannot be updates due to imports.
+
+Luckily, we can enable both capabilities. By default, when synthesis-time resolution is used, the CDK will automatically add a resource metadata entry to the template with an `Fn::ImportValue`. This will create the strong coupling between the stacks. Users can opt-out of this behavior by setting `weak: true` when they import the resource.
+
+### Convenience Methods
+
+At the top layer, we will implement a bunch of convenience methods for each AWS resource will provide nice ergonomics for cross-app import/export:
+
+Here's the usage:
+
+```ts
+// this will export "myAwesomeBucket" under the export name "JohnnyBucket"
+myAwesomeBucket.exportBucket('JohnnyBucket');
+
+// now, any CDK app that wishes to refer to this bucket can do this:
+const importedBucket: IBucket = Bucket.importBucket(this, 'JohnnyBucket', {
+  weak: true // weak-reference
+});
+```
+
+> The reason we call this `importBucket` (and `exportBucket`) is because `import` is a reserved word in Java ([#89](https://github.com/awslabs/aws-cdk/issues/89)). Also, in JavaScript both static and instance methods are inherited, so if someone extends `Bucket` (say `ExtraBucket`), we should have a way to distinguish between `ExtraBucket.importBucket` and `ExtraBucket.importExtraBucket`.
+
+The implementation of these two methods is ~trivial:
+
+```ts
+class Bucket {
+  public static importBucket(scope: Construct, exportName: string, options?: ImportOptions): IBucket {
+    return Bucket.deserializeBucket(Stack.find(scope).importObject(exportName, options));
```

Comment:
Surfacing a `stack` property on `IConstruct` should be a pre-requisite to this IMO. `Stack.find` boilerplate everywhere seems not right.